tests.py

```diff
@@ -1,129 +1,523 @@
+        * stop()            - stop server via LSP
+    """
+
+    def __init__(self, process, root_uri):
+        self.process = process
+        self.root_uri = root_uri
+        self.lsp_client = lsp.Client(
+            root_uri=root_uri,
+            workspace_folders=[lsp.WorkspaceFolder(uri=self.root_uri, name="Root")],
+            trace="verbose",
+        )
+        self.msgs = []
+
+        self._pout = process.stdout
+        self._pin = process.stdin
+
+        self._read_q = queue.Queue()
+        self._send_q = queue.Queue()
+
+        self.reader_thread = threading.Thread(
+            target=self._read_loop, name="lsp-reader", daemon=True
+        )
+        self.writer_thread = threading.Thread(
+            target=self._send_loop, name="lsp-writer", daemon=True
+        )
+
+        self.reader_thread.start()
+        self.writer_thread.start()
+
+        self.exception = None
+
+    @property
+    def all_msgs(self):
+        return self.msgs[:]
+
+    # threaded
+    def _read_loop(self):
+        try:
+            while self._pout:
+                headers, header_bytes = parse_headers(self._pout)  # type: ignore
+
+                LOG_IN and print(f"\n <<< received: {headers}\n")
+
+                if header_bytes == b"":
+                    break
+
+                body = self._pout.read(int(headers.get("Content-Length")))
+                LOG_IN and print(f"   < received: {body}\n")
+                self._read_q.put(header_bytes + body)
+        except Exception as ex:
+            self.exception = ex
+        self._send_q.put_nowait(None)  # stop send_loop()
+
+    # threaded
+    def _send_loop(self):
+        try:
+            while self._pin:
+                buf = self._send_q.get()
+
+                if buf is None:
+                    break
+
+                LOG_OUT and print(f"\nsending: {buf}\n")
+
+                self._pin.write(buf)
+                self._pin.flush()
+        except Exception as ex:
+            self.exception = ex
+
+    def _queue_data_to_send(self):
+        send_buf = self.lsp_client.send()
+        if send_buf:
+            self._send_q.put(send_buf)
+
+    def _read_data_received(self):
+        while not self._read_q.empty():
+            data = self._read_q.get()
+            events = self.lsp_client.recv(data)
+            for ev in events:
+                self.msgs.append(ev)
+                self._try_default_reply(ev)
+
+    def _try_default_reply(self, msg):
+        empty_reply_classes = (
+            lsp.ShowMessageRequest,
+            lsp.WorkDoneProgressCreate,
+            lsp.RegisterCapabilityRequest,
+            lsp.ConfigurationRequest,
+        )
+
+        if isinstance(msg, empty_reply_classes):
+            msg.reply()
+
+        elif isinstance(msg, lsp.WorkspaceFolders):
+            msg.reply([lsp.WorkspaceFolder(uri=self.root_uri, name="Root")])
+
+        else:
+            print(f"Cant autoreply: {type(msg)}")
```

Comment:
I don't think `print` shows up on terminal by default in pytest? Don't you need `--capture=no` or something?

Reply:
It does if the test fails